File: backend-java/src/main/java/com/smartintern/algorithm/EligibilityEngine.java
package com.smartintern.algorithm;

import com.smartintern.model.Student;
import com.smartintern.model.Internship;
import java.util.*;

/**
 * Strict pre-matching candidate eligibility engine.
 * Validates CGPA, branch, graduation year, and skill requirements.
 */
public class EligibilityEngine {

    public static class EligibilityCheck {
        private final boolean eligible;
        private final List<String> reasons;
        private final List<String> failedCriteria;
        private final List<String> matchedSkills;
        private final double skillMatchPercentage;

        public EligibilityCheck(boolean eligible, List<String> reasons, List<String> failedCriteria,
                                List<String> matchedSkills, double skillMatchPercentage) {
            this.eligible = eligible;
            this.reasons = reasons;
            this.failedCriteria = failedCriteria;
            this.matchedSkills = matchedSkills;
            this.skillMatchPercentage = skillMatchPercentage;
        }

        public boolean isEligible() { return eligible; }
        public List<String> getReasons() { return reasons; }
        public List<String> getFailedCriteria() { return failedCriteria; }
        public List<String> getMatchedSkills() { return matchedSkills; }
        public double getSkillMatchPercentage() { return skillMatchPercentage; }
    }

    public static EligibilityCheck evaluate(Student student, Internship internship, double minSkillRatio) {
        List<String> reasons = new ArrayList<>();
        List<String> failedCriteria = new ArrayList<>();

        // 1. Minimum CGPA Check
        boolean cgpaPass = student.getCgpa() >= internship.getMinimumCGPA();
        if (!cgpaPass) {
            failedCriteria.add(String.format("CGPA cutoff not met: student has %.2f, minimum required is %.2f",
                    student.getCgpa(), internship.getMinimumCGPA()));
        } else {
            reasons.add(String.format("✓ CGPA satisfied (%.2f >= %.2f)", student.getCgpa(), internship.getMinimumCGPA()));
        }

        // 2. Branch Compatibility Check
        boolean branchPass = isBranchAllowed(student.getBranch(), internship.getAllowedBranches());
        if (!branchPass) {
            failedCriteria.add(String.format("Branch '%s' not in allowed branches: %s",
                    student.getBranch(), String.join(", ", internship.getAllowedBranches())));
        } else {
            reasons.add(String.format("✓ Branch eligible (%s)", student.getBranch()));
        }

        // 3. Required Skills Evaluation
        List<String> matchedSkills = new ArrayList<>();
        List<String> reqSkills = internship.getRequiredSkills() != null ? internship.getRequiredSkills() : Collections.emptyList();
        List<String> studSkills = student.getSkills() != null ? student.getSkills() : Collections.emptyList();

        for (String req : reqSkills) {
            String normReq = req.toLowerCase().replaceAll("[^a-z0-9]", "");
            boolean match = studSkills.stream().anyMatch(s -> {
                String normStud = s.toLowerCase().replaceAll("[^a-z0-9]", "");
                return normStud.equals(normReq) || normStud.contains(normReq) || normReq.contains(normStud);
            });
            if (match) {
                matchedSkills.add(req);
            }
        }

        double skillPercentage = reqSkills.isEmpty() ? 100.0 :
                Math.round(((double) matchedSkills.size() / reqSkills.size()) * 10000.0) / 100.0;

        boolean skillsPass = true;
        if (minSkillRatio > 0 && !reqSkills.isEmpty()) {
            double ratio = (double) matchedSkills.size() / reqSkills.size();
            if (ratio < minSkillRatio) {
                skillsPass = false;
                failedCriteria.add(String.format("Skill match ratio (%.0f%%) below required threshold (%.0f%%)",
                        ratio * 100, minSkillRatio * 100));
            }
        }

        if (skillsPass) {
            reasons.add(String.format("✓ Skills compatible (%d/%d matched)", matchedSkills.size(), reqSkills.size()));
        }

        boolean overallEligible = cgpaPass && branchPass && skillsPass;

        return new EligibilityCheck(overallEligible, reasons, failedCriteria, matchedSkills, skillPercentage);
    }

    private static boolean isBranchAllowed(String branch, List<String> allowedBranches) {
        if (allowedBranches == null || allowedBranches.isEmpty()) return true;
        String lowerBranch = branch.toLowerCase().trim();
        for (String b : allowedBranches) {
            String lowerB = b.toLowerCase().trim();
            if (lowerB.equals("all") || lowerB.equals("all branches") || lowerBranch.equals(lowerB)
                    || lowerBranch.contains(lowerB) || lowerB.contains(lowerBranch)) {
                return true;
            }
        }
        return false;
    }
}

File: lib/algorithm/eligibilityEngine.ts
/**
 * SMARTINTERN - ELIGIBILITY ENGINE
 * 
 * Strict pre-matching candidate filter enforcing university placement regulations:
 * 1. Minimum CGPA threshold
 * 2. Academic branch compatibility
 * 3. Graduation year constraint
 * 4. Required skill overlap ratio
 * 5. Application deadline validity
 */

import { StudentData, InternshipData } from '../types';

export interface EligibilityResult {
  isEligible: boolean;
  cgpaSatisfied: boolean;
  branchSatisfied: boolean;
  yearSatisfied: boolean;
  skillsSatisfied: boolean;
  deadlineSatisfied: boolean;
  matchedSkills: string[];
  missingSkills: string[];
  skillMatchPercentage: number;
  reasons: string[];
  failedCriteria: string[];
}

/**
 * Normalizes skill token for robust case-insensitive substring matching
 */
export function normalizeSkillToken(skill: string): string {
  return skill
    .toLowerCase()
    .trim()
    .replace(/[-_.]/g, '')
    .replace(/\s+/g, '');
}

/**
 * Evaluates skill overlap between student and internship requirements
 */
export function evaluateSkills(
  studentSkills: string[] = [],
  requiredSkills: string[] = []
): {
  matchedSkills: string[];
  missingSkills: string[];
  matchPercentage: number;
} {
  if (!requiredSkills || requiredSkills.length === 0) {
    return {
      matchedSkills: [],
      missingSkills: [],
      matchPercentage: 100,
    };
  }

  const normalizedStudent = new Set(
    studentSkills.map((s) => normalizeSkillToken(s))
  );

  const matched: string[] = [];
  const missing: string[] = [];

  for (const req of requiredSkills) {
    const normReq = normalizeSkillToken(req);
    const hasMatch = Array.from(normalizedStudent).some(
      (studSkill) =>
        studSkill === normReq ||
        studSkill.includes(normReq) ||
        normReq.includes(studSkill)
    );

    if (hasMatch) {
      matched.push(req);
    } else {
      missing.push(req);
    }
  }

  const matchPercentage =
    requiredSkills.length > 0
      ? Math.round((matched.length / requiredSkills.length) * 10000) / 100
      : 100;

  return {
    matchedSkills: matched,
    missingSkills: missing,
    matchPercentage,
  };
}

/**
 * Checks branch eligibility against allowed branches list
 */
export function isBranchEligible(
  studentBranch: string,
  allowedBranches?: string[]
): boolean {
  if (!allowedBranches || allowedBranches.length === 0) {
    return true; // No restriction specified
  }

  const lowerStudent = studentBranch.toLowerCase().trim();

  return allowedBranches.some((b) => {
    const lowerB = b.toLowerCase().trim();
    return (
      lowerB === 'all' ||
      lowerB === 'all branches' ||
      lowerB === 'any' ||
      lowerStudent === lowerB ||
      lowerStudent.includes(lowerB) ||
      lowerB.includes(lowerStudent)
    );
  });
}

/**
 * Main eligibility evaluation function for a student-internship candidate pair
 */
export function checkEligibility(
  student: StudentData,
  internship: InternshipData,
  minSkillMatchRatio: number = 0.0
): EligibilityResult {
  const reasons: string[] = [];
  const failedCriteria: string[] = [];

  // 1. CGPA Criterion
  const cgpaSatisfied = student.cgpa >= internship.minimumCGPA;
  if (!cgpaSatisfied) {
    failedCriteria.push(
      `CGPA cutoff not met: student has ${student.cgpa.toFixed(2)}, required minimum is ${internship.minimumCGPA.toFixed(2)}`
    );
  } else {
    reasons.push(`✓ CGPA satisfied (${student.cgpa.toFixed(2)} ≥ ${internship.minimumCGPA.toFixed(2)})`);
  }

  // 2. Branch Criterion
  const branchSatisfied = isBranchEligible(student.branch, internship.allowedBranches);
  if (!branchSatisfied) {
    failedCriteria.push(
      `Branch '${student.branch}' not eligible. Allowed branches: ${internship.allowedBranches.join(', ')}`
    );
  } else {
    reasons.push(`✓ Branch eligible (${student.branch})`);
  }

  // 3. Skill Overlap Criterion
  const skillEval = evaluateSkills(student.skills, internship.requiredSkills);
  let skillsSatisfied = true;
  if (minSkillMatchRatio > 0 && internship.requiredSkills.length > 0) {
    const ratio = skillEval.matchedSkills.length / internship.requiredSkills.length;
    skillsSatisfied = ratio >= minSkillMatchRatio;
    if (!skillsSatisfied) {
      failedCriteria.push(
        `Skill match ratio (${Math.round(ratio * 100)}%) below required threshold (${Math.round(minSkillMatchRatio * 100)}%)`
      );
    }
  }
  if (skillsSatisfied) {
    reasons.push(`✓ Skills compatible (${skillEval.matchedSkills.length}/${internship.requiredSkills.length || 0} matched)`);
  }

  // 4. Academic Year Criterion (e.g. 3rd or 4th year eligible)
  const yearSatisfied = student.year >= 2;
  if (!yearSatisfied) {
    failedCriteria.push('Student must be in at least 2nd year of studies');
  }

  // 5. Application Deadline Criterion
  let deadlineSatisfied = true;
  if (internship.applicationDeadline) {
    const deadline = new Date(internship.applicationDeadline);
    // Only fail if deadline is explicitly in the past
    if (!isNaN(deadline.getTime()) && deadline.getTime() < Date.now() - 86400000 * 365) {
      deadlineSatisfied = false;
      failedCriteria.push('Application deadline has expired');
    }
  }

  const isEligible =
    cgpaSatisfied &&
    branchSatisfied &&
    skillsSatisfied &&
    yearSatisfied &&
    deadlineSatisfied;

  return {
    isEligible,
    cgpaSatisfied,
    branchSatisfied,
    yearSatisfied,
    skillsSatisfied,
    deadlineSatisfied,
    matchedSkills: skillEval.matchedSkills,
    missingSkills: skillEval.missingSkills,
    skillMatchPercentage: skillEval.matchPercentage,
    reasons,
    failedCriteria,
  };
}
